### PR TITLE
fix: Allow network policy based on signal type

### PR DESCRIPTION
**Pull request by rakesh-garimella**
## Description

Changes proposed in this pull request (what was done and why):

- The allow list of ports in network policy should be based on signal type for agents and gateways
- Allow scraping from istio envoy ports as well.

Changes refer to particular issues, PRs or documents:

- https://github.com/kyma-project/telemetry-manager/issues/614

## Traceability
- [x] The PR is linked to a GitHub issue.
- [ ] New features have a milestone set.
- [ ] New features have defined acceptance criteria in a corresponding GitHub Issue, and all criteria are satisfied with this PR.
- [x] The corresponding GitHub issue has a respective `area` and `kind` label.
- [ ] The follow-up issues (if any) are linked in the `Related Issues` section.
- [ ] Adjusted the documentation if the change is user-facing.
- [x] The feature is unit-tested
- [x] The feature is e2e-tested

### CODE REVIEW

File: internal/reconciler/metricpipeline/reconciler.go [hunk 1]
```diff
@@ -238,3 +244,18 @@ func (r *Reconciler) getReplicaCountFromTelemetry(ctx context.Context) int32 {
 	}
 	return defaultReplicaCount
 }
+
+func getAllowedPorts(istioActive bool) []int32 {
```
Comment (chrkl) [inline internal/reconciler/metricpipeline/reconciler.go:248]: Agent and gateway ports should be different.

**Comment by rakesh-garimella** [comment]
tested with cur to the podl by running a busybox container in k3d cluster and it works fine
```
curl -v http://xx.xx.xx.xx:15090/stats/prometheus
```

File: internal/resources/otelcollector/config.go [hunk 2]
```diff
@@ -92,3 +103,19 @@ type DaemonSetConfig struct {
...
+
+func (cfg *AgentConfig) WithAllowedPorts(ports []int32) *AgentConfig {
+	cfgCopy := *cfg
+
+	cfgCopy.allowedPorts = makePorts(ports)
+	return &cfgCopy
+
+}
+
+func makePorts(ports []int32) []intstr.IntOrString {
```
Comment (chrkl) [inline internal/resources/otelcollector/config.go:115]: Maybe let's do this conversion in `otelcollector.makeNetworkPolicyPorts`. There we already iterate over the ports and then we don't have to populate this weirt intstr type to this package.

File: internal/resources/otelcollector/core.go [hunk 3]
```diff
@@ -129,7 +129,7 @@ func makeMetricsService(name types.NamespacedName) *corev1.Service {
 	}
 }
 
-func makeDenyPprofNetworkPolicy(name types.NamespacedName) *networkingv1.NetworkPolicy {
+func makeDenyPprofNetworkPolicy(name types.NamespacedName, allowedPorts []intstr.IntOrString) *networkingv1.NetworkPolicy {
```
Comment (chrkl) [inline internal/resources/otelcollector/core.go:132]: Let's make the function name more generic (even though, the policy still has pprof in the name).

File: test/testkit/k8s/network_policy.go [hunk 4]
```diff
@@ -0,0 +1,35 @@
+package k8s
+
+import (
+	networkingv1 "k8s.io/api/networking/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type NetworkPolicy struct {
```
Comment (chrkl) [inline test/testkit/k8s/network_policy.go:8]: Let's make this name more specific. Maybe `DenyAllNetworkPolicy`?

File: internal/resources/otelcollector/core.go (hunk 3 above)
Comment (rakesh-garimella) [reply internal/resources/otelcollector/core.go]: done

File: internal/resources/otelcollector/config.go (hunk 2 above)
Comment (rakesh-garimella) [reply internal/resources/otelcollector/config.go]: done

File: internal/reconciler/metricpipeline/reconciler.go (hunk 1 above)
Comment (rakesh-garimella) [reply internal/reconciler/metricpipeline/reconciler.go]: done

File: test/testkit/k8s/network_policy.go (hunk 4 above)
Comment (rakesh-garimella) [reply test/testkit/k8s/network_policy.go]: done

**Review by chrkl — APPROVED** [review]

**Pull request merged**